chia/full_node/weight_proof_v2.py

```diff
@@ -0,0 +1,1837 @@
+    sp_info = None
+    if sub_slot_data.cc_signage_point:
+        is_overflow = is_overflow_block(constants, sub_slot_data.signage_point_index)
+        sp_iters = calculate_sp_iters(constants, ssi, sub_slot_data.signage_point_index)
+        sp_challenge = challenge
+        assert sub_slot_data.cc_sp_vdf_output
+        cc_sp_input = ClassgroupElement.get_default_element()
+        if is_overflow:
+            assert prev_challenge
+            sp_challenge = prev_challenge
+        if ssd_idx > 0 and not sub_slot_data.cc_signage_point.normalized_to_identity:
+            tmp_input, sp_iters = sub_slot_data_vdf_info(ssd_idx, sub_slots, is_overflow, ssi, sp_iters)
+            if isinstance(tmp_input, B):
+                cc_sp_input = long_outputs[tmp_input]
+            elif isinstance(tmp_input, ClassgroupElement):
+                cc_sp_input = tmp_input
+        sp_valid, sp_output = verify_compressed_vdf(
+            constants,
+            sp_challenge,
+            cc_sp_input,
+            sub_slot_data.cc_sp_vdf_output,
+            sub_slot_data.cc_signage_point,
+            sp_iters,
+        )
+        if not sp_valid:
+            raise Exception(f"failed cc signage point vdf validation {ssd_idx}")
+        long_outputs[sub_slot_data.cc_sp_vdf_output] = sp_output
+        sp_iters = calculate_sp_iters(
+            constants,
+            ssi,
+            sub_slot_data.signage_point_index,
+        )
+        assert sp_challenge
+        sp_info = VDFInfo(sp_challenge, sp_iters, sp_output)
+
+    cc_ip_input = ClassgroupElement.get_default_element()
+    ip_vdf_iters = sub_slot_data.ip_iters
+    assert sub_slot_data.cc_infusion_point
+    if not sub_slot_data.cc_infusion_point.normalized_to_identity:
+        if prev_ssd is not None and not prev_ssd.is_end_of_slot():
+            assert prev_ssd.cc_ip_vdf_output
+            cc_ip_input = long_outputs[prev_ssd.cc_ip_vdf_output]
+            assert sub_slot_data
+            assert sub_slot_data.total_iters
+            assert prev_ssd.total_iters
+            ip_vdf_iters = uint64(sub_slot_data.total_iters - prev_ssd.total_iters)
+    assert ip_vdf_iters
+    assert sub_slot_data.cc_ip_vdf_output
+    ip_valid, ip_output = verify_compressed_vdf(
+        constants,
+        challenge,
+        cc_ip_input,
+        sub_slot_data.cc_ip_vdf_output,
+        sub_slot_data.cc_infusion_point,
+        ip_vdf_iters,
+    )
+    if not ip_valid:
+        raise Exception(f"failed cc infusion point vdf validation {ssd_idx}")
+    long_outputs[sub_slot_data.cc_ip_vdf_output] = ip_output
+
+    pospace_challenge = challenge
+    assert sub_slot_data.signage_point_index is not None
+    if sampled:
+        if is_overflow_block(constants, sub_slot_data.signage_point_index):
+            assert prev_challenge
+            pospace_challenge = prev_challenge
+        __validate_pospace(constants, sub_slot_data, curr_difficulty, pospace_challenge, ssi, long_outputs)
```

Comment:
Shouldn't we also check that the iterations returned from the POSpace are correct? Anyone can make a Pospace with a trillion iterations

Reply:
good point i missed that, added